packages/wdio-cucumber-framework/src/index.ts

```diff
@@ -116,8 +128,122 @@ class CucumberAdapter {
         this._cucumberReporter = new CucumberReporter(this._eventBroadcaster, this._pickleFilter, reporterOptions, this._cid, this._specs, this._reporter)
     }
 
+    readFiles(filePaths: Options.Testrunner['specs'] = []): (string | string[])[] {
+        return filePaths.map(filePath => {
+            return Array.isArray(filePath)
+                ? filePath.map(file => fs.readFileSync(path.resolve(file), 'utf8'))
+                : fs.readFileSync(path.resolve(filePath), 'utf8')
+        })
+    }
+
+    getGherkinDocuments(files: Options.Testrunner['specs'] = []): (GherkinDocument | GherkinDocument[])[] {
+        return this.readFiles(files).map((specContent, idx) => {
+
+            const docs: GherkinDocument[] = [specContent]
+                .flat(1)
+                .map((content, ctIdx) => (
+                    {
+                        ...gherkinParser.parse(content),
+                        uri: Array.isArray(specContent)
+                            ? files[idx][ctIdx]
+                            : files[idx],
+                    } as GherkinDocument
+                ))
+
+            const [doc, ...etc] = docs
+
+            return etc.length ? docs : doc
+        })
+    }
+
+    filterSpecsByTagExpression(
+        specs: Options.Testrunner['specs'] = [],
+        tagExpression = this._config.cucumberOpts?.tagExpression ?? ''
+    ): typeof specs {
+
+        if (!tagExpression) {
+            return specs
+        }
+
+        const tagParser = TagExpressionParser(tagExpression)
+
+        const shouldRun = (doc: GherkinDocument) => {
```

Comment:
Can me move this into `utils.ts` file, this way it makes it easier to individually test the function.